proposals/extension_get_contexts.md

```diff
@@ -0,0 +1,171 @@
+
+## Solution
+
+Considering the above situation, we’d like to propose a new (Manifest V3 only)
+method to asynchronously provide metadata about the view that is still useful
+for an extension: `extension.getContexts()`. This will allow extension service
+workers (or other contexts) to identify the active contexts in the extension.
+For example, this can be used to target messages to send using
+<code>[runtime.sendMessage()](https://developer.chrome.com/docs/extensions/reference/runtime/#method-sendMessage)</code>,
+etc.). Introducing this API will allow an easier migration from MV2.
+
+## API Proposal
+
+This new API would use a similar parameter signature as
+`extension.getViews()`but instead accept a new `contextProperties` object.
+
+```
+extension.getContexts()(
+  contextProperties?: object,
+)
+```
+
+`contextProperties` will have similar behavior to `extension.getViews()`’s
+<code>[fetchProperties](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)</code>
+object in that it will limit the scope of the query to certain tabs, context
+types, and windows. It would be defined as:
+
+```
+extension.ContextProperties = {
+  tabId? = int,                // Find context by tab id, omitted returns all tabs.
+  contextType? = ContextType,  // Find context by type, omitted returns all types.
+  windowId? = int,             // Find context by window id, omitted returns all
+                               // windows.
+}
+```
+
+`ContextType` would serve a similar purpose of
+<code>[fetchProperties](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)</code>’s
+[ViewType](https://developer.chrome.com/docs/extensions/reference/extension/#type-ViewType)
+in order to limit the type of extension contexts returned by the API. It would
+be defined as:
+
+```
+extension.ContextType = {
+  TAB: 'TAB',                               // Tabs the extension is running in.
+  POPUP: 'POPUP',                           // Popups created by the extension.
+  SERVICE_WORKER: 'SERVICE_WORKER',         // Service workers for the extension.
+  OFFSCREEN_DOCUMENT: 'OFFSCREEN_DOCUMENT'  // Offscreen documents for the extension.
+};
+```
+
+It would return an array of new `Context` objects for each frame the extension
+is running in:
+
+```
+extension.Context = {
+  tabId: int,                    // ID of the tab for this view.
+  windowId: int,                 // ID of the window this tab is located in.
+  documentId: int,               // ID of the DOM document loaded in the tab.
+  frameId: int,                  // ID of the frame this context is running for.
+  contextType: ContextType,      // Context type -- tab, popup, etc.
+  url: string,                   // The current URL of the tab.
+  origin: string,                // The origin of the context. Mostly this is the
+                                 // extension origin.
+  incognito: boolean             // Whether the context is for an incognito
```

Comment:
Can we get away from browser marketing names in the API? `ephemeral` is a good neutral term for this. Documentation can tie it to the browser's marketing name.

Reply:
I'm hesitant about using `ephemeral`. While it's technically accurate (private browsing sessions are ephemerial), that label doesn't feel sufficiently specific. What about `private`? I think that better represents how users think about – and how browser describe – the feature.

Reply:
`private` works for me, does that work for you too @xeenon?

Reply:
private lgtm.

Reply:
Yes, `private` is good.

Reply:
That makes sense @Rob--W, I am fine with sticking with `incognito` to keep the API consistent.

Reply:
Agree Ron, we should be consistent, hence incognito is good.

Compared to _ephemeral_ inPrivate looks better.